kafka/source/pkg/adapter/message.go

```diff
@@ -161,3 +176,79 @@ func getKeyTypeMapper(keyType string) func([]byte) interface{} {
 	}
 	return keyTypeMapper
 }
+
+func getAvroDecoder(schemaURL string) *AvroDecoder {
+	if schemaURL != "" {
+		return &AvroDecoder{
+			schemaURL: schemaURL,
+			codecMap:  make(map[int]*goavro.Codec),
+		}
+	}
+	return nil
+}
+
+func (decoder *AvroDecoder) getCodec(schemaID int) (*goavro.Codec, error) {
+	decoder.mutex.RLock()
+	if decoder.codecMap[schemaID] != nil {
+		return decoder.codecMap[schemaID], nil
+	}
+	decoder.mutex.RUnlock()
+
+	resp, err := nethttp.Get(fmt.Sprintf("%s/schemas/ids/%d", decoder.schemaURL, schemaID))
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var schemaResp map[string]interface{}
+	if err := json.Unmarshal(body, &schemaResp); err != nil {
+		return nil, err
+	}
+
+	schema, err := json.Marshal(schemaResp["schema"])
+	if err != nil {
+		return nil, err
+	}
+
+	codec, err := goavro.NewCodec(string(schema))
+	if err != nil {
+		return nil, err
+	}
+
+	decoder.mutex.Lock()
+	decoder.codecMap[schemaID] = codec
+	decoder.mutex.Unlock()
+	return codec, nil
+}
+
+func (decoder *AvroDecoder) decode(msg *protocolkafka.Message, logger *zap.Logger) {
+	schemaID := binary.BigEndian.Uint32(msg.Value[1:5])
+	if schemaID == 0 {
+		logger.Warn("Error getting the schema id", zap.Int("schema id", int(schemaID)))
+		return
+	}
+	codec, err := decoder.getCodec(int(schemaID))
+	if err != nil {
+		logger.Warn("Error getting the schema", zap.Error(err))
+		return
+	}
+	native, _, err := codec.NativeFromBinary(msg.Value[5:])
+	if err != nil {
+		logger.Warn("Error decoding message", zap.Error(err))
```

Comment:
I think this and the one below should be decoding error. This `decode` should return an error that is propagated as `ConsumerMessageToHttpRequest` error